LDAP Connectivity Fixes

jlehtoranta opened the pull request:
This fixes the inconsistent behaviour of the user_ldap app, when a Start TLS negotiation fails. This might also (partly) fix issue #5237.

Currently, for example a Nextcloud login failure can result into a server error like this:
1. An LDAP server is configured to localhost with Start TLS enabled
2. A user tries to login to Nextcloud with a valid LDAP username, but an incorrect password
3. The bind function is called in Connection.php
4. In establishConnection first the main LDAP server is tried: Now, if the Start TLS negotiation fails, an LDAP error code is registered. The bind after that doesn't result into a registered error due to the recursion fix 86d72b9a61f5e8a9b57c6f0bb431eb6722aa12a3.
5. This leads to establishConnection trying the backup LDAP server, whether it's configured or not. The recursion fix 86d72b9a61f5e8a9b57c6f0bb431eb6722aa12a3 doesn't catch a second bind attempt anymore, so ldapConnectionRes is set to null after a failing bind.
6. Finally getConnectionResource throws a ServerNotAvailableException.

What's fixed in this PR:
- LDAP_INVALID_CREDENTIALS isn't handled as a connection error (resource not set to null)
- The backup LDAP server error handling is improved and connecting to it is tried only if it's configured
- Start TLS negotiation failures throw a ServerNotAvailableException, which prevents the app from sending plain-text bind requests
    - Check out also PR #5476, which adds the SSL/TLS status information to the LDAP settings wizard. It would be beneficial for the server admins to see that information after Start TLS failures throw an exception. Apparently some old configs might have ldap_tls enabled by default.

jlehtoranta committed:
LDAP: Connect to backup server only if it exists + handle errors

Signed-off-by: Jarkko Lehtoranta <devel@jlranta.com>

jlehtoranta committed:
LDAP: Don't handle invalid credentials as a connection error

Signed-off-by: Jarkko Lehtoranta <devel@jlranta.com>

jlehtoranta committed:
LDAP: Remove unnecessary "recursion" fix

This reverts commit 86d72b9a61f5e8a9b57c6f0bb431eb6722aa12a3
"LDAP: fix possible recursion".

Signed-off-by: Jarkko Lehtoranta <devel@jlranta.com>

jlehtoranta committed:
LDAP: Throw an exception if Start TLS fails

This ensures that only a secure connection to the LDAP server will be used,
if Start TLS has been enabled.

Signed-off-by: Jarkko Lehtoranta <devel@jlranta.com>

jlehtoranta committed:
LDAP: Fix testUseBackupServer unit test

Signed-off-by: Jarkko Lehtoranta <devel@jlranta.com>

blizzz commented on apps/user_ldap/lib/Connection.php:
```diff
@@ -550,19 +550,27 @@ private function establishConnection() {
+					}
 				}
+			} catch (\OC\ServerNotAvailableException $e) {
+				throw $e;
```
this try-catch-block is unnecessary when it's only purpose is to throw the caught extension again

jlehtoranta replied:
True, removed in the updated PR.

blizzz commented on apps/user_ldap/lib/Connection.php:
```diff
@@ -550,19 +550,27 @@ private function establishConnection() {
-					//skip contacting him until next cache refresh
-					$this->writeToCache('overrideMainServer', true);
+			try {
+				if( trim($this->configuration->ldapBackupHost) !== ""
```
This conditition is unncessary, it was checked already in line 547. If empty, we would not have reached this line. Correct me if I am wrong.

jlehtoranta replied:
In my opinion this is needed, because ServerNotAvailableException won't be thrown, if bind fails on line 539. We also cannot make bind throw a ServerNotAvailableException after an LDAP_INVALID_CREDENTIALS error, so I think it's just simplest to check the ldapBackupHost condition here too.

blizzz replied:
Ah, good point, you're right.

blizzz commented on apps/user_ldap/lib/Connection.php:
```diff
@@ -582,7 +590,9 @@ private function doConnect($host, $port) {
 				if($this->configuration->ldapTLS) {
-					$this->ldap->startTls($this->ldapConnectionRes);
+					if(!$this->ldap->startTls($this->ldapConnectionRes)) {
+						throw new \OC\ServerNotAvailableException('Start TLS failed, when connecting to LDAP host ' . $host . '.');
```
to have this work reliably we also need handle ``if($this->ldap->setOption($this->ldapConnectionRes, LDAP_OPT_REFERRALS, 0))``. When this fails, establishing startTLS won't ever be attempted.

jlehtoranta replied:
Fixed in the updated PR by throwing a ServerNotAvailableException (3b169f7).

blizzz replied:
Another thing, more cosmetic: ``\OC\ServerNotAvailableException`` is already declared in the use statement at the top of the file, just write the name ``ServerNotAvailableException``, not the whole namespace.

Additionally, instead of the nested ``if``s , use a single one and ``&&`` connect the conditions. It will be better readable and less repetitive.

blizzz reviewed (changes requested):
Thank you for this PR, for diving into this and figuring out the changes! 🚀 

I found two issues that aren't big deals, and there is another case that should be handled. If you may have a look?

Thanks again :)

jlehtoranta committed:
LDAP: Throw an exception if disabling LDAP referrals fails

Signed-off-by: Jarkko Lehtoranta <devel@jlranta.com>

blizzz commented on apps/user_ldap/lib/Connection.php:
```diff
@@ -550,19 +550,23 @@ private function establishConnection() {
 			//if LDAP server is not reachable, try the Backup (Replica!) Server
-			if(    $error !== 0
+			if( trim($this->configuration->ldapBackupHost) !== ""
+				&& ($error !== 0
```
Then the indentation is not quite right. The ``&&`` should be on the same column as the condition before is. The subsequent ones are correct.

jlehtoranta replied:
Unfortunately the Github viewer seems to default to tab width 8. However, if we're using tab width 4 that's suggested in the [Nextcloud coding guidelines](https://docs.nextcloud.com/server/12/developer_manual/general/codingguidelines.html#coding), then && is actually aligned at the same level with trim(. The lines inside the parentheses (555..556) are indented on the same level with the lines above like suggested in [Pear guidelines](http://pear.php.net/manual/en/standards.control.php). Should I align these by using tab width 8 instead of 4? Should I also add an extra tab to lines 555..556?

blizzz replied:
Yeah, however it does not make sense syntactically, and that's the bigger issue. Then, just move the ``trim(…`` into  a line of its own. :)

blizzz reviewed (commented):
Do you think you could add a unit test, that would fail without your changes?

jlehtoranta committed:
LDAP: Clean-up doConnect

Signed-off-by: Jarkko Lehtoranta <devel@jlranta.com>

jlehtoranta committed:
LDAP: Use imported exception in Connection class

Signed-off-by: Jarkko Lehtoranta <devel@jlranta.com>

jlehtoranta committed:
LDAP: Add testBindWithInvalidCredentials unit test

Signed-off-by: Jarkko Lehtoranta <devel@jlranta.com>

jlehtoranta committed:
LDAP: Add testStartTlsNegotiationFailure unit test

Signed-off-by: Jarkko Lehtoranta <devel@jlranta.com>

jlehtoranta commented:
Updated PR:
- Cleaned up the unnecessary nesting in doConnect. I used separate ifs though, since I preferred the more precise error reporting.
- Switched to the imported ServerNotAvailableException in the whole Connection class.
- Added a unit test for testing bind with invalid credentials and another for testing a Start TLS negotiation failure.

jlehtoranta committed:
LDAP: Simplify conditions in establishConnection

Signed-off-by: Jarkko Lehtoranta <devel@jlranta.com>

jlehtoranta commented:
Updated PR:

- Used two variables to simplify the lengthy two times used conditions in establishConnection

blizzz commented:
Code looks good and works. Thanks for the contribution @jlehtoranta 

need a second reviewer @nextcloud/ldap
